pkg/gardenlet/controller/vpaevictionrequirements/reconciler.go

```diff
@@ -0,0 +1,175 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package vpaevictionrequirements
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	vpaautoscalingv1 "k8s.io/autoscaler/vertical-pod-autoscaler/pkg/apis/autoscaling.k8s.io/v1"
+	"k8s.io/utils/clock"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
+	"github.com/gardener/gardener/pkg/controllerutils"
+	"github.com/gardener/gardener/pkg/gardenlet/apis/config"
+	"github.com/gardener/gardener/pkg/utils/timewindow"
+)
+
+var cpuAndMemory = []corev1.ResourceName{corev1.ResourceMemory, corev1.ResourceCPU}
+var upscaleOnlyRequirement = []*vpaautoscalingv1.EvictionRequirement{
+	{
+		Resources:         cpuAndMemory,
+		ChangeRequirement: vpaautoscalingv1.TargetHigherThanRequests,
+	},
+}
+
+func removeAllEvictionRequirements(vpa *vpaautoscalingv1.VerticalPodAutoscaler) {
+	vpa.Spec.UpdatePolicy.EvictionRequirements = []*vpaautoscalingv1.EvictionRequirement{}
+}
+
+func addDenyDownscalingEvictionRequirement(vpa *vpaautoscalingv1.VerticalPodAutoscaler) {
+	vpa.Spec.UpdatePolicy.EvictionRequirements = upscaleOnlyRequirement
+}
+
+// Reconciler implements the reconciliation logic for adding/removing EvictionRequirements to VPA objects.
+type Reconciler struct {
+	Config     config.VPAEvictionRequirementsControllerConfiguration
+	SeedClient client.Client
+	Clock      clock.Clock
+}
+
+// Reconcile implements the reconciliation logic for adding/removing EvictionRequirements to VPA objects.
+func (r *Reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	log := logf.FromContext(ctx)
+	seedCtx, cancel := controllerutils.GetMainReconciliationContext(ctx, controllerutils.DefaultReconciliationTimeout)
+	defer cancel()
+
+	vpa := &vpaautoscalingv1.VerticalPodAutoscaler{}
+	err := r.SeedClient.Get(seedCtx, request.NamespacedName, vpa)
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("error retrieving object from store: %w", err)
+	}
+
+	log.Info("Reconciling vpa")
+
+	// double check just for fun: does the vpa have our label?
+	if !metav1.HasLabel(vpa.ObjectMeta, constants.LabelVPAEvictionRequirementDownscaleRestriction) {
+		err := fmt.Errorf("label %q not found, although marker label %q is present", constants.LabelVPAEvictionRequirementDownscaleRestriction, constants.LabelVPAEvictionRequirementsController)
+		log.Error(err, "Error while parsing the label value:")
+		return reconcile.Result{}, err
+	}
```

Comment:
Nit: newline for readability

```suggestion
	}

```